packages/babel-plugin-optimize-react/index.js

```diff
@@ -0,0 +1,297 @@
+
+          if (
+            t.isCallExpression(init) &&
+            t.isIdentifier(init.node.callee) &&
+            init.node.callee.name === 'require' &&
+            init.node.arguments.length === 1 &&
+            t.isStringLiteral(init.node.arguments[0]) &&
+            init.node.arguments[0].value === 'react'
+          ) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  function isReferencingReactHook(path) {
+    if (t.isIdentifier(path)) {
+      const identifierName = path.node.name;
+      const binding = path.scope.getBinding(identifierName);
+
+      if (binding !== undefined) {
+        const bindingPath = binding.path;
+
+        if (t.isVariableDeclarator(bindingPath)) {
+          const init = bindingPath.get('init');
+          const bindingId = binding.identifier;
+
+          if (t.isIdentifier(init) && isReactImport(init)) {
+            if (reactHooks.has(bindingId.name)) {
+              return true;
+            }
+            const id = bindingPath.get('id');
+
+            if (t.isObjectPattern(id)) {
+              const properties = id.get('properties');
+
+              for (let property of properties) {
+                if (
+                  t.isObjectProperty(property) &&
+                  property.node.value === bindingId &&
+                  t.isIdentifier(property.node.key) &&
+                  reactHooks.has(property.node.key.name)
+                ) {
+                  return true;
+                }
+              }
+            }
+          } else if (t.isMemberExpression(init)) {
+            const object = init.get('object');
+            const property = init.get('property');
+
+            if (
+              isReactImport(object) &&
+              t.isIdentifier(property) &&
+              reactHooks.has(property.node.name)
+            ) {
+              return true;
+            }
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  function isUsingDestructuredArray(path) {
+    const parentPath = path.parentPath;
+
+    if (t.isVariableDeclarator(parentPath)) {
+      const id = parentPath.get('id');
+      return t.isArrayPattern(id);
+    }
+    return false;
+  }
+
+  function isCreateReactElementCall(path) {
+    if (t.isCallExpression(path)) {
+      const callee = path.get('callee');
+
+      if (t.isMemberExpression(callee)) {
+        const object = callee.get('object');
+        const property = callee.get('property');
+
+        if (
+          isReactImport(object) &&
+          t.isIdentifier(property) &&
+          property.node.name === 'createElement'
+        ) {
+          return true;
+        }
+      }
+    }
+  }
+
+  function createConstantCreateElementReference(reactReferencePath) {
+    const identifierName = reactReferencePath.node.name;
+    const binding = reactReferencePath.scope.getBinding(identifierName);
+    const createElementReference = t.identifier('__reactCreateElement__');
```

Comment:
this optimization is OK, but it works on per module (file) basis - it doesnt take into account that production bundles use scope hoisting, maybe we could somehow end up with single constant reference per chunk instead of many?

Reply:
Don't hardcode the identifier name, use path.scope.generateUidIdentifier in Program and save its value on state (the second argument to a visitor, which also is === this)

Reply:
I guess having a hardcoded name actually does let what Andarist suggested work, but not in a way that works with a minifier.

Reply:
We don't enable scope hoisting in CRA. Tbh I think it's reasonable compromise for now.

Reply:
> We don't enable scope hoisting in CRA. Tbh I think it's reasonable compromise for now.

Are you sure? I haven't checked it, but it's enabled by default in webpack@4 in production mode (which is used by CRA). https://webpack.js.org/plugins/module-concatenation-plugin/

> That would be good, but I was unsure how to do that?

Me neither 😅 Would have to move this rewrite to other phase than transpilation.